canvas/image.go

```diff
@@ -261,97 +263,97 @@ func (i *Image) updateReader() error {
 	return nil
 }
 
-func (i *Image) calculateMinSize() error {
-	var size fyne.Size
-
+func (i *Image) updateAspectAndMinSize() error {
 	if i.reader != nil {
-		var err error
-
-		size, err = i.minSizeFromReader(i.reader)
+		width, height, aspect, err := i.imageDetailsFromReader(i.reader)
 		if err != nil {
 			return err
 		}
+		i.aspect = aspect
+
+		if i.FillMode == ImageFillOriginal {
+			return i.setMinSizeForPixels(width, height)
+		}
 	} else if i.Image != nil {
 		original := i.Image.Bounds().Size()
-		size = fyne.NewSize(float32(original.X), float32(original.Y))
-		i.aspect = size.Width / size.Height
+		i.aspect = float32(original.X) / float32(original.Y)
+		if i.FillMode == ImageFillOriginal {
```

Comment:
This looks like something that could be refactored between the two if path.

Reply:
Short of moving the if into the method and calling it `setMinSizeForPixelsIfOriginalFill` I'm not sure how?
I did consider it but this layout seemed to be clearer.

Reply:
Looking at the code path both `if i.FillMode == ImageFillOriginal` seems to be doing the same things and are the last thing done in the `if`. There is also no other code path that could go past this three branch if. This automatically trigger a question for me, why is the code looking similar and duplicated, if there is no reason for it, it is misleading and duplicated. Otherwise if this code is not duplicated, then it is confusing.

Reply:
I'm not certain I follow - you mean that the code should be refactored so the `if` only occurs once? 
To do so we have to increase the scope of the width/height variables and move the code that uses the pixel information outside of the context where it is created...

```go
func (i *Image) updateAspectAndMinSize() error {
+       var pixWidth, pixHeight int
+
        if i.reader != nil {
                width, height, aspect, err := i.imageDetailsFromReader(i.reader)
                if err != nil {
                        return err
                }
                i.aspect = aspect
-
-               if i.FillMode == ImageFillOriginal {
-                       return i.setMinSizeForPixels(width, height)
-               }
+               pixWidth, pixHeight = width, height
        } else if i.Image != nil {
                original := i.Image.Bounds().Size()
                i.aspect = float32(original.X) / float32(original.Y)
-               if i.FillMode == ImageFillOriginal {
-                       return i.setMinSizeForPixels(original.X, original.Y)
-               }
+               pixWidth, pixHeight = original.X, original.Y
        } else {
                return errors.New("no matching image source")
        }
 
+       if i.FillMode == ImageFillOriginal {
+               return i.setMinSizeForPixels(pixWidth, pixHeight)
+       }
        return nil
 }
```

I personally don't see that this is preferable, but if it is required to land this tidy-up then I will add it.

Reply:
I would prefer something like that, indeed.